### PR TITLE
[cuda][eager] Preserve internal precision for native_group_norm

### PR DESCRIPTION
**Pull request by benjaminglass1**
Stack from [ghstack](https://github.com/ezyang/ghstack/tree/0.14.0) (oldest at bottom):
* #184791
* #182907
* #183980
* __->__ #183946
* #183001

The CUDA implementation for native_group_norm promotes (b)float16 inputs to float32 during computation, for accuracy.  However, we previously truncated the calculated mean and variance back to (b)float16 mid-computation.  This commit postpones truncation until the end of the kernel, increasing accuracy (and removing differences with the decomposed kernel in Inductor).

Additionally, I enabled clang-format linting for this file, which changed less than 10 lines and seemed good to enable while I was in it.

### CODE REVIEW

**Review by eqy — COMMENTED** [review]
I assume there are no substantial performance differences?

**Comment by benjaminglass1** [comment]
@eqy I'll get some benchmarks on an H100 in the cloud and report back shortly.

**Comment by benjaminglass1** [comment]
@eqy Ended up only being able to get access to an A100 rather than an H100, but here are the results for a bfloat16 run I did (I've tweaked the implementation so that no actual extra allocation occurs for non-accelerated/higher bitwidth types, so they should be completely unaffected).

```
[---------------------- dtype=torch.bfloat16 ------------------------------]
                                                           |    Old |    New
22 threads: ----------------------------------------------------------------
      input shape (13, 18, 32, 32), without weight/bias    |     21 |     25
      input shape (13, 18, 32, 32), with weight/bias       |     29 |     34
      input shape (13, 18, 127, 127), without weight/bias  |     44 |     43
      input shape (13, 18, 127, 127), with weight/bias     |     46 |     46
      input shape (13, 18, 512, 512), without weight/bias  |   1119 |   1117
      input shape (13, 18, 512, 512), with weight/bias     |   1119 |   1120
      input shape (13, 48, 32, 32), without weight/bias    |     21 |     26
      input shape (13, 48, 32, 32), with weight/bias       |     30 |     33
      input shape (13, 48, 127, 127), without weight/bias  |    188 |    187
      input shape (13, 48, 127, 127), with weight/bias     |    190 |    190
      input shape (13, 48, 512, 512), without weight/bias  |   2961 |   2957
      input shape (13, 48, 512, 512), with weight/bias     |   2956 |   2956
      input shape (13, 72, 32, 32), without weight/bias    |     22 |     26
      input shape (13, 72, 32, 32), with weight/bias       |     30 |     33
      input shape (13, 72, 127, 127), without weight/bias  |    284 |    283
      input shape (13, 72, 127, 127), with weight/bias     |    287 |    287
      input shape (13, 72, 512, 512), without weight/bias  |   4441 |   4436
      input shape (13, 72, 512, 512), with weight/bias     |   4427 |   4427
      input shape (64, 18, 32, 32), without weight/bias    |     22 |     26
      input shape (64, 18, 32, 32), with weight/bias       |     30 |     34
      input shape (64, 18, 127, 127), without weight/bias  |    160 |    159
      input shape (64, 18, 127, 127), with weight/bias     |    162 |    162
      input shape (64, 18, 512, 512), without weight/bias  |   2417 |   2411
      input shape (64, 18, 512, 512), with weight/bias     |   2397 |   2396
      input shape (64, 48, 32, 32), without weight/bias    |     30 |     29
      input shape (64, 48, 32, 32), with weight/bias       |     33 |     34
      input shape (64, 48, 127, 127), without weight/bias  |    405 |    404
      input shape (64, 48, 127, 127), with weight/bias     |    404 |    404
      input shape (64, 48, 512, 512), without weight/bias  |   6425 |   6407
      input shape (64, 48, 512, 512), with weight/bias     |   6350 |   6350
      input shape (64, 72, 32, 32), without weight/bias    |     42 |     41
      input shape (64, 72, 32, 32), with weight/bias       |     44 |     44
      input shape (64, 72, 127, 127), without weight/bias  |    598 |    596
      input shape (64, 72, 127, 127), with weight/bias     |    598 |    598
      input shape (64, 72, 512, 512), without weight/bias  |   9640 |   9609
      input shape (64, 72, 512, 512), with weight/bias     |   9555 |   9553

Times are in microseconds (us).
```

As you can see, time differentials appear to be mostly noise, if anything. Benchmarking script is below, although I had to manually collate the table (since the two runs are on separate compiled builds).

```
from functools import partial
from itertools import product

import torch
from torch.testing import make_tensor
from torch.utils.benchmark import Compare, Timer

benchmark_name = "native_group_norm"
device = "cuda:0"

dtypes = (
    torch.bfloat16
    if torch.cuda.get_device_capability(device) >= (8, 0)
    else torch.float16,
    torch.float32,
    torch.float64,
)
batch_sizes = (13, 64)
channel_multipliers = (3, 8, 12)
input_sizes = (32, 127, 512)


def gen_inputs():
    for dtype, batch_size, channel_mul, AB, with_weight_bias in product(
        dtypes, batch_sizes, channel_multipliers, input_sizes, (False, True)
    ):
        num_groups = 6
        num_channels = num_groups * channel_mul
        make_arg = partial(make_tensor, dtype=dtype, device=device)

        input_shape = (batch_size, num_channels, AB, AB)
        input = make_arg(input_shape)

        weight = make_arg(num_channels) if with_weight_bias else None
        bias = make_arg(num_channels) if with_weight_bias else None

        yield input, weight, bias, batch_size, num_channels, AB * AB, num_groups, 0.0314


def benchmark(label, f, *args):
    dtype = args[0].dtype

    # Warm up the timed function kernel.
    f(*args)
    f(*args)

    return Timer(
        "f(*args)",
        globals={"f": f, "args": args},
        description=label,
        label=f"{dtype=}",
        sub_label=f"input shape {tuple(args[0].shape)}, {'with' if args[1] is not None else 'without'} weight/bias",
        num_threads=torch.get_num_threads(),
    ).adaptive_autorange()


def compare(*args):
    f = torch.ops.aten.native_group_norm.default
    yield benchmark("Eager", f, *args)


c = Compare([r for args in gen_inputs() for r in compare(*args)])
c.trim_significant_figures()
c.print()
```

**Review by eqy — APPROVED** [review]

**Comment by benjaminglass1** [comment]
@pytorchbot merge -i

Failures are unrelated
